// @jetbrains/ring-ui/components/contenteditable/contenteditable.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Component } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
/**
 * @name ContentEditable
 */
function noop() { }
class ContentEditableBase extends Component {
    static defaultProps = {
        disabled: false,
        tabIndex: 0,
        onInput: noop,
        onComponentUpdate: noop,
    };
    shouldComponentUpdate(nextProps) {
        return nextProps.disabled !== this.props.disabled || nextProps.__html !== this.props.__html;
    }
    componentDidUpdate(prevProps) {
        this.props.onComponentUpdate(prevProps);
    }
    render() {
        const { __html, onComponentUpdate, disabled, tabIndex, inputRef, ...props } = this.props;
        return (_jsx("div", { ...props, ref: inputRef, 
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            disabled: disabled, role: "textbox", tabIndex: disabled ? undefined : tabIndex, contentEditable: !this.props.disabled, dangerouslySetInnerHTML: { __html } }));
    }
}
const ContentEditable = ({ children, ...props }) => (_jsx(ContentEditableBase, { ...props, __html: renderToStaticMarkup(children) }));
export default ContentEditable;
