// @jetbrains/ring-ui/components/heading/heading.js
import { jsx as _jsx } from "react/jsx-runtime";
import { memo } from 'react';
import classNames from 'classnames';
import deprecate from 'util-deprecate';
import styles from './heading.css';
/**
 * @name Heading
 */
export var Levels;
(function (Levels) {
    Levels[Levels["H1"] = 1] = "H1";
    Levels[Levels["H2"] = 2] = "H2";
    Levels[Levels["H3"] = 3] = "H3";
    Levels[Levels["H4"] = 4] = "H4";
})(Levels || (Levels = {}));
const fallbackHeading = deprecate(() => 'h3', 'Headings of level 5 and higher are replaced with h3');
const HeadingMemo = memo(function Heading({ children, className, level = Levels.H1, ...restProps }) {
    const classes = classNames(styles.heading, className);
    const Tag = level <= Levels.H4 ? `h${level}` : fallbackHeading();
    return (_jsx(Tag, { ...restProps, className: classes, children: children }));
});
const Heading = HeadingMemo;
Heading.Levels = Levels;
export default Heading;
const H1 = memo(function H1(props) {
    return _jsx(Heading, { ...props, level: Levels.H1 });
});
const H2 = memo(function H2(props) {
    return _jsx(Heading, { ...props, level: Levels.H2 });
});
const H3 = memo(function H3(props) {
    return _jsx(Heading, { ...props, level: Levels.H3 });
});
const H4 = memo(function H4({ className, bold, ...restProps }) {
    const classes = classNames(className, { [styles.bold]: bold });
    return _jsx(Heading, { ...restProps, className: classes, level: Levels.H4 });
});
export { H1, H2, H3, H4 };
