// @jetbrains/ring-ui/components/date-picker/consts.js
import { add } from 'date-fns/add';
const unit = 8; // px;
const units = {
    unit,
    /* eslint-disable @typescript-eslint/no-magic-numbers */
    cellSize: unit * 3,
    calHeight: unit * 36,
    yearHeight: unit * 4,
    /* eslint-enable */
};
export default units;
export const YEAR = 12;
export const WEEK = 7;
export const FIFTH_DAY = 4;
export const weekdays = {
    MO: 1,
    TU: 2,
    WE: 3,
    TH: 4,
    FR: 5,
    SA: 6,
    SU: 0,
};
export const MIDDLE_DAY = 15;
const durationToMillis = (duration) => +add(0, duration);
export const yearDuration = durationToMillis({ years: 1 });
export const yearScrollSpeed = yearDuration / (YEAR * units.cellSize);
export const DOUBLE = 2;
export const HALF = 0.5;
export function parseTime(time) {
    let result = null;
    if (/^([01][0-9]|2[0-3]):[0-5][0-9]$/.test(time)) {
        result = time;
    }
    else if (/^([0-9]|2[0-3]):[0-5][0-9]$/.test(time)) {
        result = `0${time}`;
    }
    return result;
}
export function shiftWeekArray(arr, startOfWeek) {
    const shiftTimes = startOfWeek - 1;
    return arr.slice(shiftTimes).concat(arr.slice(0, shiftTimes));
}
export function getWeekStartsOn(locale) {
    return locale?.options?.weekStartsOn ?? weekdays.MO;
}
export function getDayNumInWeek(locale, day) {
    const weekDays = shiftWeekArray(Object.values(weekdays), getWeekStartsOn(locale));
    return weekDays.indexOf(day);
}
