// @jetbrains/ring-ui/components/user-card/tooltip.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Component } from 'react';
import classNames from 'classnames';
import Dropdown from '../dropdown/dropdown';
import Popup from '../popup/popup';
import UserCard from './card';
import styles from './user-card.css';
const DEFAULT_TIMEOUT = 300;
export default class UserCardTooltip extends Component {
    static defaultProps = {
        renderUserCard: (props) => {
            const { user, ...restProps } = props;
            return user && _jsx(UserCard, { user: user, ...restProps });
        },
        renderNoUser: () => '',
        dropdownProps: {
            hoverShowTimeOut: DEFAULT_TIMEOUT,
            hoverHideTimeOut: DEFAULT_TIMEOUT,
        },
    };
    render() {
        const { children, renderUserCard, renderNoUser, dropdownProps, user, ...restProps } = this.props;
        return (_jsx(Dropdown, { anchor: children, hoverMode: true, clickMode: false, ...dropdownProps, children: _jsx(Popup, { attached: false, children: user
                    ? renderUserCard({
                        ...restProps,
                        user,
                        className: classNames(styles.userCardSpaced, this.props.className),
                    })
                    : renderNoUser() }) }));
    }
}
