// @jetbrains/ring-ui/components/footer/footer.js
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * @name Footer
 */
import { memo, isValidElement } from 'react';
import classNames from 'classnames';
import Link from '../link/link';
import styles from './footer.css';
const FooterColumn = memo(function FooterColumn({ position, children }) {
    const classes = classNames({
        [styles.columnLeft]: position === 'left',
        [styles.columnCenter]: position === 'center',
        [styles.columnRight]: position === 'right',
    });
    return (_jsx("div", { className: classes, children: _jsx("ul", { className: styles.columnItem, children: children }) }));
});
/**
 * Return copyright string
 * @param year {int}
 * @returns {string}
 */
export function copyright(year) {
    const currentYear = new Date().getUTCFullYear();
    const ndash = '–';
    let ret = 'Copyright © ';
    if (year >= currentYear) {
        ret += year;
    }
    else {
        ret += year + ndash + currentYear;
    }
    return ret;
}
/**
 * @constructor
 * @extends {ReactComponent}
 */
const FooterLine = memo(function FooterLine(props) {
    const items = Array.isArray(props.item) ? props.item : [props.item];
    function renderItem(item) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        if (isValidElement(item) || typeof item !== 'object') {
            return item;
        }
        const element = (item.copyright ? copyright(item.copyright) : '') + (item.label || item);
        if (item.url) {
            return (_jsx(Link, { href: item.url, target: item.target, title: item.title, children: element }, item.url + item.title));
        }
        return element;
    }
    return _jsx("li", { className: styles.line, children: items.map(renderItem) });
});
const Footer = memo(function Footer({ floating, className, left, center, right }) {
    function content(elements, position) {
        if (!elements) {
            return false;
        }
        return (_jsx(FooterColumn, { position: position, children: elements.map((item, idx) => (_jsx(FooterLine
            // eslint-disable-next-line react/no-array-index-key
            , { item: item }, idx))) }, position));
    }
    const classes = classNames(styles.footer, className, {
        [styles.footerFloating]: floating,
    });
    return (_jsx("footer", { className: classes, "data-test": "ring-footer", children: [content(left, 'left'), content(center, 'center'), content(right, 'right')] }));
});
export default Footer;
