// @jetbrains/ring-ui/components/grid/col.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Component } from 'react';
import classNames from 'classnames';
import styles from './grid.css';
const classMap = {
    xs: 'col-xs',
    sm: 'col-sm',
    md: 'col-md',
    lg: 'col-lg',
    xsOffset: 'col-xs-offset',
    smOffset: 'col-sm-offset',
    mdOffset: 'col-md-offset',
    lgOffset: 'col-lg-offset',
};
/**
 * Converts props like "xs=9 xsOffset={2}" to classes "col-xs-9 col-xs-offset-2"
 * @param {Object} props incoming props
 * @returns {Array} result classnames
 */
function getClassNames(props) {
    return Object.entries(props)
        .filter(([key]) => classMap[key])
        .map(([key, value]) => styles[Number.isInteger(value) ? `${classMap[key]}-${value}` : classMap[key]]);
}
export default class Col extends Component {
    render() {
        const { children, className, reverse, ...restProps } = this.props;
        const classes = classNames(styles.col, className, getClassNames(restProps), {
            [styles.reverse]: reverse,
        });
        return (_jsx("div", { className: classes, "data-test": "ring-grid-column", children: children }));
    }
}
