// @jetbrains/ring-ui/components/popup-menu/popup-menu.js
import { jsx as _jsx } from "react/jsx-runtime";
import Popup from '../popup/popup';
import List from '../list/list';
/**
 * @name Popup Menu
 */
export default class PopupMenu extends Popup {
    static isItemType = List.isItemType;
    static ListProps = List.ListProps;
    static defaultProps = {
        ...List.defaultProps,
        ...Popup.defaultProps,
        renderOptimization: false,
        closeOnSelect: false,
        largeBorderRadius: true,
    };
    onSelect = (item, event) => {
        if (this.props.closeOnSelect) {
            this._onCloseAttempt(event);
        }
        this.props.onSelect(item, event);
    };
    list;
    listRef = (el) => {
        this.list = el;
    };
    /** @override */
    getInternalContent() {
        const { className, ...props } = this.props;
        return (_jsx("div", { children: _jsx(List, { ref: this.listRef, ...props, maxHeight: this.popup && parseFloat(this.popup.style.maxHeight), shortcuts: this.shouldUseShortcuts(), onSelect: this.onSelect }) }));
    }
}
export const { ListProps } = List;
