// @jetbrains/ring-ui/components/table/smart-table.js
import { jsx as _jsx } from "react/jsx-runtime";
import { PureComponent } from 'react';
import Table from './table';
import Selection from './selection';
class SmartTable extends PureComponent {
    static defaultProps = {
        onSelectionChange: () => { },
    };
    state = {
        selection: new Selection({
            data: this.props.data,
            isItemSelectable: this.props.isItemSelectable,
        }),
    };
    UNSAFE_componentWillReceiveProps(nextProps) {
        const { data, isItemSelectable, selection } = nextProps;
        if (this.props.remoteSelection && this.props.selection !== selection) {
            this.setState({ selection });
        }
        else if (this.props.data !== data || this.props.isItemSelectable !== isItemSelectable) {
            this.setState({
                selection: new Selection({ data, isItemSelectable }),
            });
        }
    }
    onSelect = (selection) => {
        this.setState({ selection });
        this.props.onSelectionChange(selection);
    };
    render() {
        return _jsx(Table, { ...this.props, selection: this.state.selection, onSelect: this.onSelect });
    }
}
export default SmartTable;
