// @jetbrains/ring-ui/components/shortcuts/shortcuts.js
import { PureComponent } from 'react';
import shortcuts from './core';
export default class Shortcuts extends PureComponent {
    static defaultProps = {
        options: {},
    };
    componentDidMount() {
        if (!this.props.disabled) {
            this.turnShorcutsOn();
        }
    }
    componentDidUpdate(prevProps) {
        const { disabled, map } = this.props;
        if (!prevProps.disabled && disabled) {
            this.turnShorcutsOff();
        }
        if (prevProps.disabled && !disabled) {
            this.turnShorcutsOn();
        }
        if (prevProps.map !== map) {
            shortcuts.bindMap(map, this.props);
        }
    }
    componentWillUnmount() {
        if (!this.props.disabled) {
            this.turnShorcutsOff();
        }
    }
    turnShorcutsOn() {
        const { map, scope, options } = this.props;
        shortcuts.bindMap(map, this.props);
        shortcuts.pushScope(scope, options);
    }
    turnShorcutsOff() {
        const { scope } = this.props;
        shortcuts.unbindScope(scope);
        shortcuts.spliceScope(scope);
    }
    render() {
        return this.props.children || null;
    }
}
