// @jetbrains/ring-ui/components/data-list/data-list.mock.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import Link from '../link/link';
import Tag from '../tag/tag';
const items = [
    {
        id: 1,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Assigner" }), " in 60 projects"] })),
        collapsible: true,
        selectable: true,
        items: [
            {
                id: 11,
                title: '6 projects: as a member of jetbrains-team',
                collapsible: true,
                selectable: true,
                items: [
                    {
                        id: 111,
                        selectable: true,
                        title: _jsx(Link, { href: "#", children: "TeamCity" }),
                    },
                    {
                        id: 112,
                        selectable: true,
                        title: _jsx(Link, { href: "#", children: "YouTrack" }),
                    },
                    {
                        id: 113,
                        selectable: true,
                        title: _jsx(Link, { href: "#", children: "Sandbox" }),
                    },
                    {
                        id: 114,
                        selectable: true,
                        title: _jsx(Link, { href: "#", children: "Exception Analyzer" }),
                    },
                    {
                        id: 115,
                        selectable: true,
                        title: _jsx(Link, { href: "#", children: "MPS" }),
                    },
                ],
            },
        ],
    },
    {
        id: 2,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Code Reviewer" }), " in 5 projects"] })),
        selectable: true,
        collapsible: true,
        items: [
            {
                id: 21,
                selectable: true,
                title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "JetProfile" }), _jsx("span", { children: " " }), _jsx(Tag, { children: "duplicate" })] })),
            },
            {
                id: 22,
                title: '6 projects: as a member of jetbrains-team',
                collapsible: true,
                items: [
                    {
                        id: 221,
                        collapsible: true,
                        title: _jsx(Link, { href: "#", children: "TeamCity" }),
                    },
                    {
                        id: 222,
                        collapsible: true,
                        title: _jsx(Link, { href: "#", children: "YouTrack" }),
                    },
                    {
                        id: 223,
                        collapsible: true,
                        title: _jsx(Link, { href: "#", children: "Sandbox" }),
                    },
                    {
                        id: 224,
                        collapsible: true,
                        title: _jsx(Link, { href: "#", children: "Exception Analyzer" }),
                    },
                    {
                        id: 225,
                        collapsible: true,
                        title: _jsx(Link, { href: "#", children: "MPS" }),
                    },
                ],
            },
            {
                id: 23,
                collapsible: true,
                title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "CustomJDK" }), _jsx("span", { children: ": as a member of ide-developers" })] })),
            },
        ],
    },
    {
        id: 3,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Code Viewer" }), " in 5 projects"] })),
        selectable: true,
        collapsible: true,
        items: [
            {
                id: 31,
                collapsible: true,
                title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "ide-frontend" }), _jsx("span", { children: ": as a member of ide-developers" })] })),
            },
            {
                id: 32,
                collapsible: true,
                title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "ide-frontend" }), _jsx("span", { children: ": as a member of ide-frontend Code Viewers" })] })),
            },
        ],
    },
    {
        id: 4,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Commenter" }), " in 12 projects"] })),
        collapsible: true,
        items: [
            {
                id: 41,
                collapsible: true,
                title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "Intellij IDEA" }), _jsx("span", { children: ": as a member of Registered Users" })] })),
            },
            {
                id: 42,
                collapsible: true,
                title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "YouTrack Backlog" }), _jsx("span", { children: ": as a member of jetbrains-team" })] })),
            },
        ],
    },
    {
        id: 5,
        collapsible: true,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Developer" }), _jsx("span", { children: " in 57 projects " }), _jsx(Tag, { children: "team role" })] })),
        items: [
            {
                id: 51,
                selectable: true,
                collapsible: true,
                title: _jsx(Link, { href: "#", children: "Wasabi" }),
            },
        ],
    },
    {
        id: 6,
        collapsible: true,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Git Hosting Admin" }), _jsx("span", { children: " in project " }), _jsx(Link, { href: "#", children: "Hackathon 2016" }), _jsx("span", { children: ": as a member of jetbrains-team" })] })),
    },
    {
        id: 7,
        title: (_jsxs("span", { children: [_jsx("strong", { children: "Git Hosting Checkout" }), " in 7 projects"] })),
        selectable: true,
        collapsible: true,
    },
];
export const moreItems = [
    {
        id: 91,
        selectable: true,
        collapsible: true,
        title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "More item 1" }), _jsx("span", { children: " " }), _jsx(Tag, { children: "duplicate" })] })),
    },
    {
        id: 92,
        selectable: true,
        collapsible: true,
        title: (_jsxs("span", { children: [_jsx(Link, { href: "#", children: "More item 2" }), _jsx("span", { children: " " }), _jsx(Tag, { children: "duplicate" })] })),
    },
];
export default items;
